### PR TITLE
BFS+GUPs benchmarks

### PR DESCRIPTION
**Pull request by mrutt92**

### CODE REVIEW

File: scripts/py/dramsim3_utils.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    with open(filename_i, "r") as jsonfile:
+        lines = [line for line in jsonfile]
+
+    for (lno, line) in enumerate(lines):
+        # if this is the last line
+        # replace the ',' with ']'
+        if lno == len(lines)-1:
+            line = line.replace(',\n',']\n')
+        # write it out
+        jsonfile_o.write(line)
```
Comment (drichmond) [inline scripts/py/dramsim3_utils.py:16]: Little confused why this is necessary

I thought we had a fix that closed out the json file. Is this the same issue?

File: machines/bigblade_pod_X1Y1_ruche_X16Y8_hbm/Makefile.machine.include [hunk 2]
```diff
@@ -0,0 +1,135 @@
+# Copyright (c) 2019, University of Washington All rights reserved.
```
Comment (drichmond) [inline machines/bigblade_pod_X1Y1_ruche_X16Y8_hbm/Makefile.machine.include:1]: I would separate this into another PR

File: libraries/features/dma/simulation/dramsim3.mk [hunk 3]
```diff
@@ -64,6 +64,7 @@ $(LIBRARIES_PATH)/features/dma/simulation/libdramsim3.so: CXXFLAGS += -I$(BASEJU
 $(LIBRARIES_PATH)/features/dma/simulation/libdramsim3.so: CXXFLAGS += -I$(BASEJUMP_STL_DIR)/imports/DRAMSim3/ext/fmt/include
 $(LIBRARIES_PATH)/features/dma/simulation/libdramsim3.so: CXXFLAGS += -DFMT_HEADER_ONLY=1
 $(LIBRARIES_PATH)/features/dma/simulation/libdramsim3.so: CXXFLAGS += -DBASEJUMP_STL_DIR="$(BASEJUMP_STL_DIR)"
+$(LIBRARIES_PATH)/features/dma/simulation/libdramsim3.so: CXXFLAGS += -DBLOOD_GRAPH=1
```
Comment (drichmond) [inline libraries/features/dma/simulation/dramsim3.mk:67]: Always on?

File: examples/cuda/gups/.gitignore [hunk 4]
```diff
@@ -0,0 +1 @@
+*_run/
```
Comment (drichmond) [inline examples/cuda/gups/.gitignore:1]: Newline plz

File: examples/cuda/bfs/.gitignore [hunk 5]
```diff
@@ -0,0 +1,6 @@
+input-analyze
+input-gen
+input-degree-list
+inputs/
+*tgy/
+stack.info.*
```
Comment (drichmond) [inline examples/cuda/bfs/.gitignore:6]: Newline plz

File: .gitmodules [hunk 6]
```diff
@@ -0,0 +1,3 @@
+[submodule "examples/cuda/gups/hammerblade-helpers"]
+	path = examples/cuda/gups/hammerblade-helpers
+	url = git@github.com:mrutt92/hammerblade-helpers
```
Comment (drichmond) [inline .gitmodules:3]: I feel like this should be somewhere else in the directory tree long-term, but since you're the only one using it right now, LGTM

File: scripts/py/vcache_utils.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
+import pandas as pd
+
```
Comment (drichmond) [inline scripts/py/vcache_utils.py:2]: If you ever want it, I have the CudaStatTag class popped out in another branch

File: scripts/py/dramsim3_utils.py (hunk 1 above)
Comment (mrutt92) [reply scripts/py/dramsim3_utils.py:16]: I was surprised too, looks like it didn't make it in to the .tag.json.

Comment (drichmond) [reply scripts/py/dramsim3_utils.py:16]: @marcusnchow Do you remember anything about this? I thought we fixed it.

Comment (marcusnchow) [reply scripts/py/dramsim3_utils.py:16]: [This](https://github.com/bsg-external/DRAMsim3/blob/1b8828cd293077e321a028310a8b938031b21ff5/src/dram_system.cc#L73) was the fix that i made that replaces the last ']' with a ','

Not sure if this is the same file that this script is working with?
Or do you know if BaseDRAMSystem::PrintStats() is being called?

File: .gitmodules (hunk 6 above)
Comment (mrutt92) [reply .gitmodules:3]: Yeah there's a PR that was never merged.

File: scripts/py/vcache_utils.py (hunk 7 above)
Comment (mrutt92) [reply scripts/py/vcache_utils.py:2]: good to know. let this be the start of a stats scripting directory in replicant? honestly, I just use It to reformat the csv into a table I can more easily manage in excel.

File: libraries/features/dma/simulation/dramsim3.mk (hunk 3 above)
Comment (mrutt92) [reply libraries/features/dma/simulation/dramsim3.mk:67]: Hrm. Probably not? We could make a version of libdramsim3 with blood graphs compiled in and then only link when profiling. That would more or less match what we do for the vanilla blood graphs right? With the tracer?

File: machines/bigblade_pod_X1Y1_ruche_X16Y8_hbm/Makefile.machine.include (hunk 2 above)
Comment (mrutt92) [reply machines/bigblade_pod_X1Y1_ruche_X16Y8_hbm/Makefile.machine.include:1]: OK, see #744

File: .gitmodules (hunk 6 above)
Comment (mrutt92) [reply .gitmodules:3]: See #724

File: machine.mk [hunk 8]
```diff
@@ -38,7 +38,7 @@ endif
 
 # To switch machines, simply switch the path of BSG_MACHINE_PATH to
 # another directory with a Makefile.machine.include file.
-BSG_MACHINE_PATH ?= $(BSG_F1_DIR)/machines/pod_X1Y1_ruche_X16Y8_hbm
+BSG_MACHINE_PATH ?= $(BSG_F1_DIR)/machines/bigblade_pod_X1Y1_ruche_X16Y8_hbm
```
Comment (drichmond) [inline machine.mk:41]: Is this intentional? I thought if it was, it would go in #744

File: libraries/features/dma/simulation/dramsim3.mk (hunk 3 above)
Comment (drichmond) [reply libraries/features/dma/simulation/dramsim3.mk:67]: Do DRAMSim3 bloodgraphs slow down sim?

File: .gitmodules (hunk 6 above)
Comment (drichmond) [reply .gitmodules:3]: Wondered. Do you want that merged, or keep this as a submodule for now?

File: scripts/py/vcache_utils.py (hunk 7 above)
Comment (drichmond) [reply scripts/py/vcache_utils.py:2]: Any scripts that are generally useful for manycore should go in manycore repository, but it seems like you're still iterating on this so I'm fine with this for now.

File: scripts/py/dramsim3_utils.py (hunk 1 above)
Comment (mrutt92) [reply scripts/py/dramsim3_utils.py:16]: OK, my bad. My helper library uses a smart pointer and it calls DPI finish functions for the device after the finally blocks in the SV. I've added calls to explicitly perform HB cleanup code.

File: machine.mk (hunk 8 above)
Comment (mrutt92) [reply machine.mk]: removed. not intentional.

File: scripts/py/vcache_utils.py (hunk 7 above)
Comment (mrutt92) [reply scripts/py/vcache_utils.py:2]: well you know what they say about rules of thumb... some are made to be broken.

Comment (mrutt92) [reply scripts/py/vcache_utils.py:2]: you could make the argument that this more generally useful for cuda than for manycore.

File: .gitmodules (hunk 6 above)
Comment (mrutt92) [reply .gitmodules:3]: I don't know. We should talk about it.

File: libraries/features/dma/simulation/dramsim3.mk (hunk 3 above)
Comment (mrutt92) [reply libraries/features/dma/simulation/dramsim3.mk:67]: I'm going to go ahead and say "probably"

File: scripts/py/vcache_utils.py (hunk 7 above)
Comment (drichmond) [reply scripts/py/vcache_utils.py:2]: Wait, thumbs are meant to be broken?

You heard the discussion in meeting this morning. Replicant is a public repo but manycore is the focus repository. If it's generally useful, and it's finalized it should go in manycore.

File: .gitmodules (hunk 6 above)
Comment (drichmond) [reply .gitmodules:3]: Urgently?

File: libraries/features/dma/simulation/dramsim3.mk (hunk 3 above)
Comment (drichmond) [reply libraries/features/dma/simulation/dramsim3.mk:67]: Then we should come up with a way of turning it on and off, unless we know otherwise

File: .gitmodules (hunk 6 above)
Comment (mrutt92) [reply .gitmodules:3]: nah, this is something we can fix later.

Comment (mrutt92) [reply .gitmodules:3]: I'm more concerned with getting this merged so it doesn't go stale.

File: scripts/py/vcache_utils.py (hunk 7 above)
Comment (mrutt92) [reply scripts/py/vcache_utils.py:2]: Yeah I heard that... but I think when it comes to data post-processing from an experiment I think most researchers are particular enough that they'll have their own flow anyway. That's why it's great that are stats infrastructure within the simulation just dumps out a CSV.

Comment (drichmond) [reply scripts/py/vcache_utils.py:2]: SGTM. Then let's move the scripts directory inside bfs or gups since it's your flow at the moment

File: scripts/py/dramsim3_utils.py (hunk 1 above)
Comment (drichmond) [reply scripts/py/dramsim3_utils.py:16]: Cool

File: scripts/py/vcache_utils.py (hunk 7 above)
Comment (mrutt92) [reply scripts/py/vcache_utils.py:2]: No. This is shared code between the two benchmarks, I'll likely reuse it for more, and I think it's OK for replicant to have a common scripts directory.

File: libraries/features/dma/simulation/dramsim3.mk (hunk 3 above)
Comment (mrutt92) [reply libraries/features/dma/simulation/dramsim3.mk:67]: Maybe I can tackle that in a separate PR. It's not super integral to this one.

File: scripts/py/vcache_utils.py (hunk 7 above)
Comment (mrutt92) [reply scripts/py/vcache_utils.py:2]: But if I can always make another submodule.

Comment (mrutt92) [reply scripts/py/vcache_utils.py]: well instead I just moved it into the one that already exists.

**Comment by mrutt92** [comment]
@drichmond your feedback has been addressed, merge?

**Review by drichmond — APPROVED** [review]

**Comment by drichmond** [comment]
Yeah, LGTM
